lib/src/Base/Geom/Mesh.cxx

```diff
@@ -555,6 +555,53 @@ Point Mesh::getUpperBound() const
   return vertices_.getMax();
 }
 
+/* Orientation management */
+void Mesh::fixOrientation()
+{
+  const UnsignedInteger size = getSimplicesNumber();
+  for (UnsignedInteger i = 0; i < size; ++i)
+    fixOrientation(i);
+}
+
+void Mesh::fixOrientation(const UnsignedInteger & index)
+{
+  if (getDimension() == 2)
+    {
+      IndicesCollection::iterator cit = simplices_.begin_at(index);
+      const Scalar v1x = vertices_(*cit, 0);
+      const Scalar v1y = vertices_(*cit, 1);
+      const Scalar v2x = vertices_(*cit + 1, 0);
+      const Scalar v2y = vertices_(*cit + 1, 1);
+      const Scalar v3x = vertices_(*cit + 2, 0);
+      const Scalar v3y = vertices_(*cit + 2, 1);
+      if ((v3x - v2x) * (v1y - v2y) < (v1x - v2x) * (v3y - v2y)) std::swap(*cit, *(cit + 1));
+      return;
+    }
+  if (getDimension() == 3)
+    {
+      IndicesCollection::iterator cit = simplices_.begin_at(index);
+      const Scalar v1x = vertices_(*cit, 0);
+      const Scalar v1y = vertices_(*cit, 1);
+      const Scalar v1z = vertices_(*cit, 2);
+      const Scalar v2x = vertices_(*cit + 1, 0);
+      const Scalar v2y = vertices_(*cit + 1, 1);
+      const Scalar v2z = vertices_(*cit + 1, 2);
+      const Scalar v3x = vertices_(*cit + 2, 0);
+      const Scalar v3y = vertices_(*cit + 2, 1);
+      const Scalar v3z = vertices_(*cit + 2, 2);
+      const Scalar v4x = vertices_(*cit + 3, 0);
+      const Scalar v4y = vertices_(*cit + 3, 1);
+      const Scalar v4z = vertices_(*cit + 3, 2);
+      if ((v1x - v4x) * ((v2y - v4y) * (v3z - v4z) - (v3y - v4y) * (v2z - v4z)) + (v3x - v4x) * ((v1y - v4y) * (v2z - v4z) - (v2y - v4y) * (v1z - v4z)) < (v2x - v4x) * ((v1y - v4y) * (v3z - v4z) - (v3y - v4y) * (v1z - v4z)) ) std::swap(*cit, *(cit + 1));
+      return;
+    }
+  SquareMatrix matrix(buildSimplexMatrix(index));
```

Comment:
In checkPointInSimplex, I inlined call to buildSimplexMatrix to avoid SquareMatrix allocation.  The same idea could be used here, but it is surely better to add SquareMatrix& argument to buildSimplexMatrix (and to fixOrientation of course).  This is uncommon, but these methods are not meant for users, and almost always called when looping over simplices.

Reply:
@dbarbier If I understand correctly your proposal, the signature of Mesh::buildSimplexMatrix() should be `void Mesh::buildSimplexMatrix(const UnsignedInteger index, SquareMatrix & matrix)`, with the caller in charge of the allocation? Maybe with a test to see if the given matrix has the correct dimension?

Reply:
Yes and yes, and this method should then not be wrapped by SWIG, and not documented.